src/cleanfigure.m

```diff
@@ -114,16 +114,32 @@ function cleanfigure(varargin)
           movePointsCloser(meta, h);
           % Don't be too precise.
           coarsenLine(meta, h, minimumPointsDistance);
+
       elseif strcmp(type, 'text')
-          % Check if text is inside bounds by checking if the Extent rectangle
-          % and the axes box overlap.
-          xlim = get(meta.gca, 'XLim');
-          ylim = get(meta.gca, 'YLim');
+          % Check if text is inside bounds by checking if the position is inside
+          % the x, y and z limits. This works for both 2D and 3D plots.
+          x_lim = get(meta.gca, 'XLim');
+          y_lim = get(meta.gca, 'YLim');
+          z_lim = get(meta.gca, 'ZLim');
+          axLim = [x_lim; y_lim; z_lim];
+
+          pos = get(h, 'Position');
+          bPosInsideLim = ( pos' >= axLim(:,1) ) & ( pos' <= axLim(:,2) );
+
+          % In 2D plots the extent of the textbox is available and also
+          % considered to keep the textbox, if it is partially inside the axis
+          % limits.
           extent = get(h, 'Extent');
-          extent(3:4) = extent(1:2) + extent(3:4);
-          overlap = xlim(1) < extent(3) && xlim(2) > extent(1) ...
-                 && ylim(1) < extent(4) && ylim(2) > extent(2);
-          if ~overlap
+          if all(~isnan(extent))
```

Comment:
It took me quite a while to understand what is going on here. So perhaps, we should put a comment to document the idea behind this.
As I see it: the X/Y boundaries are extended to the top left to make sure that e.g. if you have a text box that starts left/above of the axes, and overlaps it partially, it gets included.

I wonder whether that's always what we want (e.g. imagine a text label placed in the middle of a 2x2 subplot).
Wouldn't it make more sense to check if the text box lies within (or intersects) the bounding box of all axes?
Anyhow: I think the current approach is an improvement, so this should be merged once it is documented a bit more.

The other thing that popped up in my mind was: is it guaranteed that `extent` is given in plot units (i.e. as XLim/YLim/...)? I can imagine that is the default situation, but I couldn't find that property in the help at first glance.